File: src/main/java/network/reflected/rfnetapi/ReflectedAPI.java
package network.reflected.rfnetapi;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import network.reflected.rfnetapi.chestgui.GUIProvider;
import network.reflected.rfnetapi.commands.CommandRegistry;
import network.reflected.rfnetapi.purchases.PurchaseAPI;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

@RequiredArgsConstructor
public class ReflectedAPI {
    @Getter private final RfnetAPI plugin;

    private static final List<APIGetCallback> callbacks = new ArrayList<>();
    static void checkCallbacks() {
        ReflectedAPI api = get();
        if (api != null) {
            for (APIGetCallback callback : callbacks) {
                Bukkit.getScheduler().runTask(api.plugin, () -> callback.run(api));
            }
            callbacks.clear();
        }
    }

    /**
     * Returns an instance of the API
     * May be null if called before this plugin loads,
     * so depend on this plugin in your plugin.yml.
     * @see ReflectedAPI#get(network.reflected.rfnetapi.ReflectedAPI.APIGetCallback)
     *
     * @return An instance of this class
     */
    @Nullable
    public static ReflectedAPI get() {
        @Nullable Plugin plugin = Bukkit.getPluginManager().getPlugin("RfnetAPI");
        if (plugin instanceof RfnetAPI) {
            return ((RfnetAPI) plugin).getApi();
        } else {
            return null;
        }
    }

    /**
     * Gives an instance of this class via a callback
     * so that your plugin doesn't have to depend on RfnetAPI.
     *
     * @param lambda The callback to run when the plugin has loaded
     */
    public static void get(APIGetCallback lambda) {
        ReflectedAPI api = get();
        if (api != null) {
            lambda.run(api);
            checkCallbacks();
        } else {
            callbacks.add(lambda);
        }
    }

    @Getter private final GUIProvider guiProvider = new GUIProvider();
    @Getter private final PurchaseAPI purchaseAPI = new PurchaseAPI();
    @Getter private final CommandRegistry commandProvider = new CommandRegistry();

    public String getLoadedMap() {
        return plugin.getLoadedMapName();
    }

    public boolean isMinigameWorldSupportEnabled() {
        return plugin.isMinigameWorld();
    }

    public void setAvailable(boolean available) {
        plugin.getDatabase().setAvailable(available);
    }

    public void restart() {
        plugin.restart();
    }

    public void sendPlayer(Player player, String archetype) {
        plugin.sendPlayer(player, archetype);
    }

    public interface APIGetCallback {
        void run(@NotNull ReflectedAPI api);
    }
}

File: src/main/java/network/reflected/rfnetapi/ReflectedAPI.kt
package network.reflected.rfnetapi

import network.reflected.rfnetapi.bugs.ExceptionDispensary
import network.reflected.rfnetapi.commands.CommandRegistry
import network.reflected.rfnetapi.purchases.PurchaseAPI
import org.bukkit.Bukkit
import org.bukkit.World
import org.bukkit.entity.Player
import org.bukkit.plugin.IllegalPluginAccessException

class ReflectedAPI internal constructor(private val plugin: RfnetAPI) {
    val purchaseAPI = PurchaseAPI()
    val commandProvider = CommandRegistry()

    fun setAvailable(available: Boolean) {
        try {
            plugin.database.setAvailable(available)
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "setting server availability")
        }
    }

    fun getVersion(): Int {
        return plugin.ver
    }

    fun restart() {
        try {
            plugin.restart()
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "restarting")
        }
    }

    val database: Database
        get() = plugin.database

    fun sendPlayer(player: Player, archetype: String) {
        try {
            plugin.sendPlayer(player, archetype)
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "sending player")
        }
    }

    fun getLoadedMap(): World {
        try {
            return plugin.loadedMap?.let {
                Bukkit.getWorld(it.name)
            } ?: Bukkit.getWorlds()[0]
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "getting loaded map")
            throw NullPointerException("Error whilst getting loaded map. This was reported to the internal exception log.")
        }
    }

    companion object { // Back compat
        fun get(): ReflectedAPI {
            return getReflectedAPI()
        }
    }
}

/**
 * Returns an instance of the API
 * May be null if called before this plugin loads,
 * so depend on this plugin in your plugin.yml.
 * @see ReflectedAPI.get
 * @return An instance of this class
 */
fun getReflectedAPI(): ReflectedAPI {
    val plugin = Bukkit.getPluginManager().getPlugin("RfnetAPI")
    if (plugin is RfnetAPI && plugin.api != null) {
        return plugin.api!!
    } else {
        throw IllegalPluginAccessException("Plugin must depend on RfnetAPI to use ReflectedAPI.get()")
    }
}